### PR TITLE
Fix bug in arm64 LazyMachState unwind

### PR DESCRIPTION
**Pull request by janvorli**
When we were enabling the mac M1 couple of years ago, a bug has creeped in that causes a race in lazy machine state unwinding. The LazyMachState::setLazyStateFromUnwind was incorrectly modifying the captured state, so in case multiple threads were racing for the unwind, a thread that came at a time when part of the captured state was modified and before the `_isValid` was set would start unwinding from a wrong state (the partial unwound one), effectively unwinding twice.

I've hit this in CI for enabling the new exception handling, as it results in an assert when comparing the resulting address of the floating point registers unwind with the mach state.

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

**Comment by janvorli** [comment]
Hmm, many macOS arm64 tests are failing with this change, so it seems that the fix is not as easy as it looked. It is interesting though that linux arm64 works just fine.

**Comment by janvorli** [comment]
Ok, I understand the issue. The fix needs to be different. On arm64 Unix (or at least macOS), we need to add the unwound register values to the MachState, like we do for amd64. The actual bug is due to the fact that we use the captured state for keeping the values of the unwound state, which causes the race.

**Comment by janvorli** [comment]
@jkotas I have updated the fix to work properly on macOS arm64 too. The way is the same as we have always used for macOS x64.

**Comment by janvorli** [comment]
> Why is the HelperMethodFrame unwinding different on macOS?

macOS libunwind doesn't support context pointers. I'll add a comment.

**Comment by janvorli** [comment]
@jkotas I've added a comment you've suggested and also unified piece of code between Linux and Windows since both always have context pointers.

File: src/coreclr/vm/arm64/gmscpu.h [hunk 1]
```diff
@@ -26,6 +26,8 @@ struct MachState {
     TADDR          _sp; // stack pointer after the function returns
     BOOL           _isValid;
 #ifdef __APPLE__
```
Comment (jkotas) [inline src/coreclr/vm/arm64/gmscpu.h:28]: We do this unconditionally for all Unix on x64: https://github.com/dotnet/runtime/blob/5e1caf14fd94b4ed85357002303c8935fedacb8b/src/coreclr/vm/amd64/gmscpu.h#L74-L77

Should it be the same?

Comment (janvorli) [reply src/coreclr/vm/arm64/gmscpu.h:28]: It should for the sake of having it equivalent on both architecture, I just didn't want to touch that code due to your plans to get rid of all the helper method frames soon. There is nothing wrong with the amd64 way, it is just useless on Linux.

Comment (janvorli) [reply src/coreclr/vm/arm64/gmscpu.h:28]: I'd be ok to make that change for x64 too though if you prefer that.

Comment (jkotas) [reply src/coreclr/vm/arm64/gmscpu.h:28]: Ok, let's hope we will get rid of it soon.

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**